### PR TITLE
feat(net): add NetworkMetrics

### LINKED ISSUE #684: Add NetworkMetrics type
### Describe the feature

similar to other Metrics types, for example https://github.com/paradigmxyz/reth/pull/584

add a `NetworkMetrics` type that contains all related counters/gauges for example for

* num connected peers
* pending sessions
* pending session failures
* num incoming connections
* total closed sessions
* ...

This should be integrated into the `NetworkManager`

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by TechieBoy**
Closes #684

### CODE REVIEW

File: docs/design/metrics.md [hunk 1]
```diff
@@ -69,6 +69,12 @@ How the metrics are exposed to the end-user is determined by the CLI.
 - `transaction_pool.invalid_transactions`: Number of invalid transactions 
 - `transaction_pool.removed_transactions`: Number of removed transactions from the pool
 
+#### Network Manager
+- `network_manager.num_connected_peers`: Number of peers currently connected to
+- `network_manager.pending_session_failures`: Cumulative number of failures of pending sessions
+- `network_manager.total_closed_sessions`: Total number of sessions closed
+- `network_manager.num_incoming_connections`: Total number of incoming connections handled
```
Comment (rkrasiuk) [inline docs/design/metrics.md:76]: the default separator is `_`, so the above metrics will be prefixed with `network_manager_`

**Review by rkrasiuk — CHANGES_REQUESTED** [review]

File: crates/net/network/src/manager.rs [hunk 2]
```diff
@@ -537,6 +541,7 @@ where
                 }
                 SwarmEvent::IncomingTcpConnection { remote_addr, session_id } => {
                     trace!(target : "net", ?session_id, ?remote_addr, "Incoming connection");
+                    this.metrics.num_incoming_connections.increment(1);
                 }
                 SwarmEvent::OutgoingTcpConnection { remote_addr, peer_id } => {
                     trace!(target : "net", ?remote_addr, ?peer_id, "Starting outbound connection.");
```
Comment (gakonst) [inline crates/net/network/src/manager.rs:547]: needs one here too

File: crates/net/network/src/metrics.rs [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    pub(crate) num_connected_peers: Gauge,
+
+    /// Cumulative number of failures of pending sessions
+    pub(crate) pending_session_failures: Counter,
+
+    /// Total number of sessions closed
+    pub(crate) total_closed_sessions: Counter,
+
+    /// Total Number of incoming connections
+    pub(crate) num_incoming_connections: Counter,
```
Comment (gakonst) [inline crates/net/network/src/metrics.rs:17]: num outgoing?

File: crates/net/network/src/manager.rs [hunk 4]
```diff
@@ -537,6 +541,7 @@ where
                 }
                 SwarmEvent::IncomingTcpConnection { remote_addr, session_id } => {
                     trace!(target : "net", ?session_id, ?remote_addr, "Incoming connection");
+                    this.metrics.num_incoming_connections.increment(1);
```
Comment (gakonst) [inline crates/net/network/src/manager.rs:544]: shouldn't there be metrics for every SwarmEvent variant?

File: docs/design/metrics.md (hunk 1 above)
Comment (mattsse) [reply docs/design/metrics.md:76]: we should drop the `_manager` part entirely, and just have `network_`

File: crates/net/network/src/manager.rs [hunk 5]
```diff
@@ -537,9 +542,11 @@ where
                 }
                 SwarmEvent::IncomingTcpConnection { remote_addr, session_id } => {
                     trace!(target : "net", ?session_id, ?remote_addr, "Incoming connection");
+                    this.metrics.incoming_connections.increment(1);
```
Comment (onbjerg) [inline crates/net/network/src/manager.rs:545]: this does not seem to be decremented ever, imo the metric is more useful if it shows you how many at any point in time are incoming vs outgoing

Comment (mattsse) [reply crates/net/network/src/manager.rs:545]: this makes sense, I guess both would be useful, `total_incoming` and `incoming`?

File: crates/net/network/src/metrics.rs [hunk 6]
```diff
@@ -0,0 +1,27 @@
+use metrics::{Counter, Gauge};
+use reth_metrics_derive::Metrics;
+
+#[derive(Metrics)]
+#[metrics(scope = "network")]
```
Comment (mattsse) [inline crates/net/network/src/metrics.rs:5]: nit: add brief docs, one-liner

File: crates/net/network/src/manager.rs (hunk 5 above)
Comment (onbjerg) [reply crates/net/network/src/manager.rs:545]: total_incoming can be derived more easily, just having one is fine

**Review by mattsse — CHANGES_REQUESTED** [review]
this looks great, only thing left is some differentiation between total/current connection types.

**Review by mattsse — APPROVED** [review]
great!

let's merge and iterate on it if we need to tune what we want to collect

**Pull request merged**